contributors/design-proposals/grow-volume-size.md

```diff
@@ -0,0 +1,310 @@
+
+1. Write a library that abstracts away various file system operations, such as - resizing, formatting etc.
+
+   Pros:
+   * Relatively well known pattern
+
+   Cons:
+   * Depending on version with which Kubernetes is compiled with, we are still tied to which file systems are supported in which version
+     of kubernetes.
+2. Ship a wrapper shell script that encapsulates various file system operations and as long as the shell script supports particular file system
+   the resize operation is supported.
+   Pros:
+   * Kubernetes Admin can easily replace default shell script with her own version and thereby adding support for more file system types.
+
+   Cons:
+   * I don't know if there is a pattern that exists in kube today for shipping shell scripts that are called out from code in Kubernetes. Flex is
+     different because, none of the flex scripts are shipped with Kuberntes.
+3. Ship resizing tools in a container.
+
+
+Of all options - #3 is our best bet but we are not quite there yet. Hence, I would like to propose that we ship with support for
+most common file systems in curent release and we revisit this coupling and solve it in next release.
+
+## API and UI Design
+
+Given a PVC definition:
+
+```yaml
+kind: PersistentVolumeClaim
+apiVersion: v1
+metadata:
+ name: volume-claim
+ annotations:
+   volume.beta.kubernetes.io/storage-class: "generalssd"
+spec:
+ accessModes:
+   - ReadWriteOnce
+ resources:
+   requests:
+     storage: 1Gi
+```
+
+Users can request new size of underlying PV by simply editing the PVC and requesting new size:
+
+```
+~> kubectl edit pvc volume-claim
+kind: PersistentVolumeClaim
+apiVersion: v1
+metadata:
+ name: volume-claim
+ annotations:
+   volume.beta.kubernetes.io/storage-class: "generalssd"
+spec:
+ accessModes:
+   - ReadWriteOnce
+ resources:
+   requests:
+     storage: 10Gi
+```
+
+## API Changes
+
+### PVC API Change
+
+`pvc.spec.resources.requests.storage` field of pvc object will become mutable after this change.
+
+In addition to that PVC's status will have a `Conditions []PvcCondition` - which will be used
+to communicate the status of PVC to the user.
+
+The API change will be protected by Alpha feature gate and api-server will not allow PVCs with
+`Status.Conditions` field if feature is not enabled. `omitempty` in serialization format will
+prevent presence of field if not set.
+
+So the `PersistentVolumeClaimStatus` will become:
+
+```go
+type PersistentVolumeClaimStatus struct {
+    Phase PersistentVolumeClaimPhase
+    AccessModes []PersistentVolumeAccessMode
+    Capacity ResourceList
+    // New Field added as part of this Change
+    Conditions []PVCCondition
+}
+
+// new API type added
+type PVCCondition struct {
+   Type PVCConditionType
+   Status ConditionStatus
+   LastProbeTime metav1.Time
+   LastTransitionTime metav1.Time
+   Reason string
+   Message string
+}
+
+// new API type
+type PVCConditionType string
+
+// new Constants
+const (
```

Comment:
@jsafrane provided a feedback based on actual implementation that, we need only one Condition not three. We just need a single `ResizeInProgress` (or called `ResizeStarted` ) condition and once resize is finished we can delete the Condition from pvc. 

All messaging to user can be provided via events rather Conditions. I don't massively disagree and we can drop other Conditions since errors are also available via events too.

Reply:
I'm ok with this